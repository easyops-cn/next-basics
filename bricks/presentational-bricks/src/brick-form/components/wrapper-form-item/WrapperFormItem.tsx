import React from "react";
import { Form } from "@ant-design/compatible";
import { FormComponentProps } from "@ant-design/compatible/lib/form";
import { FieldDefinition } from "../../BrickForm";
import { BrickOptionProps } from "../../interfaces";
import { forEach } from "lodash";
import rewritePattern from "regexpu-core";

export interface WrapperFormItemProps
  extends FieldDefinition,
    FormComponentProps {
  allFields: FieldDefinition[];
  optionList: string[] & BrickOptionProps[];
  onFieldChange: (value: any, field: string) => void;
}

export function WrapperFormItem(WrapperComponent: any) {
  return function WithWrapperItem(props: WrapperFormItemProps) {
    const {
      isRequire,
      label,
      form,
      field,
      allFields,
      hideFromField,
      optionList,
      defaultValue,
      valuePropName = "value",
      onFieldChange,
      rules,
      ...restProps
    } = props;
    let itemShow = true;
    const overwriteProps: Record<string, any> = {};

    overwriteProps.onChange = (value: any) => {
      onFieldChange && onFieldChange(value, field);
    };

    // 根据其他表单项的值控制该表单隐藏和显示
    if (hideFromField !== undefined) {
      const targetField = allFields.find(
        (item: FieldDefinition) => item.field === hideFromField
      ).defaultValue;
      const value = form.getFieldValue(hideFromField);
      itemShow = !!(value !== undefined ? value : targetField);
    }

    if (optionList !== undefined) {
      overwriteProps.optionList = optionList.map((option) =>
        typeof option === "string" ? { id: option, text: option } : option
      );
    }

    // storyboard 中无法写正则表达式，所以写成字符串这里再进行转化。
    // storyboard 中配置的正则表达式字符串需要进行双重转义，例如希望匹配/\.at/，则写成"\\.at"，规则同jest
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp#Description
    if (rules) {
      forEach(rules, (rule) => {
        if (rule.pattern && typeof rule.pattern === "string") {
          const rewriteRulePattern = rewritePattern(rule.pattern, rule.flags, {
            unicodePropertyEscape: true,
            useUnicodeFlag: false,
          });
          rule.pattern = new RegExp(rewriteRulePattern, rule.flags);
          delete rule.flags;
        }
      });
    }

    return (
      itemShow && (
        <Form.Item label={label}>
          {form.getFieldDecorator(field, {
            initialValue: defaultValue,
            rules: rules || [],
            valuePropName,
          })(<WrapperComponent {...restProps} {...overwriteProps} />)}
        </Form.Item>
      )
    );
  };
}
