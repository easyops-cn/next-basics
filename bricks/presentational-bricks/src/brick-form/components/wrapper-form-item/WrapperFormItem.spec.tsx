import React from "react";
import { shallow } from "enzyme";
import { WrapperFormItem } from "./WrapperFormItem";

describe("WrapperFormItem", () => {
  const props = {
    field: "name",
    label: "名称",
    form: {
      getFieldDecorator: () => (comp: React.Component) => comp
    },
    onFieldChange: jest.fn()
  };
  const TestComponent = () => <div>should works</div>;

  it("should work", () => {
    const WrapperComponent = WrapperFormItem(TestComponent);
    const wrapper = shallow(
      <WrapperComponent {...props} optionList={["a", "b", "c"]} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should trigger change event", () => {
    const onChange = jest.fn();
    const WrapperComponent = WrapperFormItem(TestComponent);
    const wrapper = shallow(
      <WrapperComponent {...props} isRequire={true} onChange={onChange} />
    );
    wrapper.find(TestComponent).invoke("onChange")();
    expect(wrapper).toMatchSnapshot();
  });

  it("should control the form item show/hidden", () => {
    const newProps = {
      field: "name",
      label: "名称",
      component: "Input",
      hideFromField: "enable",
      rules: [
        {
          pattern: "^[a-zA-Z0-9]{1,100}$",
          message: "产品名称只能包含中英文、数字和_，且不能超过100个字符"
        }
      ],
      allFields: [
        {
          field: "name",
          label: "名称",
          component: "Input",
          hideFromField: "enable"
        },
        {
          field: "enable",
          label: "开启",
          component: "Switch",
          valuePropName: "checked",
          defaultValue: false
        }
      ],
      form: {
        getFieldDecorator: () => (comp: React.Component) => comp,
        getFieldValue: jest.fn().mockReturnValue(undefined)
      },
      onChange: jest.fn()
    };
    const WrapperComponent = WrapperFormItem(TestComponent);
    const wrapper = shallow(<WrapperComponent {...newProps} />);

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({
      form: {
        getFieldDecorator: () => (comp: React.Component) => comp,
        getFieldValue: jest.fn().mockReturnValue(true)
      }
    });

    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });
});
