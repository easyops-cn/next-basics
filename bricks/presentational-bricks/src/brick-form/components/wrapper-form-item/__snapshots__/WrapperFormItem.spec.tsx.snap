// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WrapperFormItem should control the form item show/hidden 1`] = `""`;

exports[`WrapperFormItem should control the form item show/hidden 2`] = `
<FormItem
  hasFeedback={false}
  label="名称"
>
  <TestComponent
    component="Input"
    onChange={[Function]}
  />
</FormItem>
`;

exports[`WrapperFormItem should trigger change event 1`] = `
<FormItem
  hasFeedback={false}
  label="名称"
>
  <TestComponent
    onChange={[Function]}
  />
</FormItem>
`;

exports[`WrapperFormItem should work 1`] = `
<FormItem
  hasFeedback={false}
  label="名称"
>
  <TestComponent
    onChange={[Function]}
    optionList={
      Array [
        Object {
          "id": "a",
          "text": "a",
        },
        Object {
          "id": "b",
          "text": "b",
        },
        Object {
          "id": "c",
          "text": "c",
        },
      ]
    }
  />
</FormItem>
`;
